### PR TITLE
Port iseq GC callbacks

### PR DESCRIPTION
**Pull request by XrXr**
rb_yjit_iseq_free() is close to the allocation site for the payload
cause they are related.

### CODE REVIEW

**Comment by XrXr** [comment]
A bit of a yak shave to add conversion methods between VALUE and `{cme,iseq}` but 🤷🏼 maybe it'll help with debugging.

File: yjit/src/core.rs [hunk 1]
```diff
@@ -377,6 +378,158 @@ fn get_iseq_payload(iseq: IseqPtr) -> &'static mut IseqPayload
     unsafe { payload_non_null.as_mut() }.unwrap()
 }
 
+/// Free the per-iseq YJTI payload
```
Comment (noahgibbs) [inline yjit/src/core.rs]: Typo: YJTI -> YJIT

File: yjit/src/core.rs [hunk 2]
```diff
@@ -377,6 +378,158 @@ fn get_iseq_payload(iseq: IseqPtr) -> &'static mut IseqPayload
...
+pub extern "C" fn rb_yjit_iseq_free(payload: *mut c_void)
+{
+    let payload = {
+        if payload.is_null() {
+            // Nothing to free.
+            return;
+        } else {
+            payload as *mut IseqPayload
+        }
+    };
```
Comment (maximecb) [inline yjit/src/core.rs:392]: Detail maybe, but you could pass an IseqPtr to these functions instead. Then you'd have access to the `get_iseq_payload` method which does the unwrapping, and you'd be able to print the iseq pointer for debugging (if ever needed).

**Review by maximecb — COMMENTED** [review]
Looks good with minor comments 👍

File: yjit/src/core.rs (hunk 2 above)
Comment (XrXr) [reply yjit/src/core.rs:392]: `get_iseq_payload` allocates the payload when it's null though so to make that change I would need to touch some other functions -- wouldn't want to be allocating and then immediately freeing the payload. I think it's a good call to make the change for other reasons but it feels out of scope for this PR.

**Review by maximecb — APPROVED** [review]

**Comment by XrXr** [comment]
CI caught an issue with RefCell borrowing so I'm looking into that. (thanks for adding the temporary CI check 👍🏼 )

```
thread '<unnamed>' panicked at 'already mutably borrowed: BorrowError', src/core.rs:430:37
```

**Comment by maximecb** [comment]
> CI caught an issue with RefCell borrowing so I'm looking into that. (thanks for adding the temporary CI check 👍🏼 )
> 
> ```
> thread '<unnamed>' panicked at 'already mutably borrowed: BorrowError', src/core.rs:430:37
> ```

I noticed that error in other failing tests too 👍 Thanks for letting me know you'll be looking into it :)

**Pull request merged**